Update main.py

bkvie opened the pull request:
Funnily enough, when putting split value=1, util.py checks for isinstance(split, float) , which fails for obvious reasons and defaults to default_split. This easily fixes that.

ClementPinard commented:
This is actually always a float if you select 1 from cli, you can see it [here](https://github.com/ClementPinard/FlowNetPytorch/blob/master/main.py#L37) I did some tests when selecting a `split-value` of 1 and got 1.0 in `args.split_value`.

Did you manually set the split_value inside the python code ?

I think the code could be robust to numerical values that are not float (here 0 or 1) but maybe we should handle this better. Here is the code I propose.

```python
def split2list(images, split, default_split=0.9):
    if isinstance(split, str):
        with open(split) as f:
            split_values = [x.strip() == '1' for x in f.readlines()]
        assert(len(images) == len(split_values))
    elif split is None:
        split_values = np.random.uniform(0,1,len(images)) < default_split
    else:
        try:
            split = float(split)
        except ValueError:
            print("Invalid Split value, it must be either a filepath or a float")
            raise
        split_values = np.random.uniform(0,1,len(images)) < split
    train_samples = [sample for sample, split in zip(images, split_values) if split]
    test_samples = [sample for sample, split in zip(images, split_values) if not split]
    return train_samples, test_samples
```

Notice how now default value is only used when split is explicetely set to None.
After that the code checks that the split can be called by the `float` function.

If you are fine with this code, you can amend your commit with it and I'll push the PR.

bkvie commented:
Awesome thanks ! Yup I  have called it from python code, great explanation sorry I couldn’t be of more help. I will amend my commit  !

Sent from my iPhone

On 12.11.2018, at 18:52, Clément Pinard <notifications@github.com<mailto:notifications@github.com>> wrote:


This is actually always a float if you select 1 from cli, you can see it here<https://github.com/ClementPinard/FlowNetPytorch/blob/master/main.py#L37> I did some tests when selecting a split-value of 1 and got 1.0 in args.split_value.

Did you manually set the split_value inside the python code ?

I think the code could be robust to numerical values that are not float (here 0 or 1) but maybe we should handle this better. Here is the code I propose.

def split2list(images, split, default_split=0.9):
    if isinstance(split, str):
        with open(split) as f:
            split_values = [x.strip() == '1' for x in f.readlines()]
        assert(len(images) == len(split_values))
    elif split is None:
        split_values = np.random.uniform(0,1,len(images)) < default_split
    else:
        try:
            split = float(split)
        except ValueError:
            print("Invalid Split value, it must be either a filepath or a float")
            raise
        split_values = np.random.uniform(0,1,len(images)) < split
    train_samples = [sample for sample, split in zip(images, split_values) if split]
    test_samples = [sample for sample, split in zip(images, split_values) if not split]
    return train_samples, test_samples

Notice how now default value is only used when split is explicetely set to None.
After that the code checks that the split can be called by the float function.

If you are fine with this code, you can amend your commit with it and I'll push the PR.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/ClementPinard/FlowNetPytorch/pull/50#issuecomment-437971998>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AH9tY2afBAAF-pByT2lJ_raHH_O1ZwLuks5uubVWgaJpZM4YZh_r>.

ClementPinard commented:
Hey I thought a bit about it and actually, when split is neither None nor a number, the error thrown will be type error and not value error. The value error is raised when the string is not a number, but in that case we would already be in the first case.

Also maybe we should manage the error the same way the correlation sampler error is managed : with a simple warning and back to default value.

see [here](https://github.com/ClementPinard/FlowNetPytorch/blob/master/models/util.py#L4-L11) for the example I mentionned.

What's your thoughts on this ?

ClementPinard commented:
Closing this as I took care of it in the main branch

bkvie commented:
Sorry was busy with a deadline, missed my chance on a pull request :/
